packages/omgidl-parser/src/IdlNodes/ReferenceTypeIdlNode.ts

```diff
@@ -0,0 +1,204 @@
+import { EnumIdlNode, IdlNode } from "./IdlNode";
+import { StructIdlNode } from "./StructIdlNode";
+import { BaseASTNode, StructMemberASTNode, TypedefASTNode } from "../astTypes";
+import { SIMPLE_TYPES, normalizeType } from "../primitiveTypes";
+import { IDLMessageDefinitionField } from "../types";
+
+type PossibleParentNode = StructIdlNode | TypedefIdlNode | EnumIdlNode;
+
+export class ReferenceTypeIdlNode<
+  T extends TypedefASTNode | StructMemberASTNode,
+> extends IdlNode<T> {
+  private needsResolution = false;
+  private parentNode?: PossibleParentNode;
+  constructor(scopePath: string[], astNode: T, idlMap: Map<string, IdlNode>) {
+    super(scopePath, astNode, idlMap);
+    // We do not protect against circular typedefs
+    if (!SIMPLE_TYPES.has(astNode.type)) {
+      this.needsResolution = true;
+    }
+  }
+
+  get type(): string {
+    if (this.needsResolution) {
+      const parent = this.parent();
+      if (parent instanceof StructIdlNode) {
+        return parent.scopedIdentifier;
+      }
+      return this.parent().type;
+    }
+    return this.astNode.type;
+  }
+
+  get isComplex(): boolean {
+    return this.needsResolution ? this.parent().isComplex : false;
+  }
+
+  get isArray(): boolean | undefined {
+    let isArray = this.astNode.isArray;
+
+    if (this.needsResolution) {
+      const parent = this.parent();
+      if (parent instanceof StructIdlNode) {
+        return this.astNode.isArray;
+      }
+      if (parent.isArray != undefined) {
+        isArray = parent.isArray;
+      }
+    }
+    return isArray;
+  }
+
+  get arrayLengths(): number[] | undefined {
+    const arrayLengths = this.astNode.arrayLengths ? [...this.astNode.arrayLengths] : [];
+    if (this.needsResolution) {
+      const parent = this.parent();
+      if (!(parent instanceof StructIdlNode) && parent.arrayLengths) {
+        arrayLengths.push(...parent.arrayLengths);
+      }
+    }
+    const finalArrayLengths: number[] = [];
+    for (const arrayLength of arrayLengths) {
+      if (typeof arrayLength === "object") {
+        finalArrayLengths.push(this.getConstantNode(arrayLength.name).value as number);
+        continue;
+      }
+      finalArrayLengths.push(arrayLength);
+    }
+    return finalArrayLengths.length > 0 ? finalArrayLengths : undefined;
+  }
+
+  get arrayUpperBound(): number | undefined {
+    let arrayUpperBound = undefined;
+    if (this.needsResolution) {
+      const parent = this.parent();
+      if (parent instanceof StructIdlNode) {
+        return undefined;
+      }
+      arrayUpperBound = parent.arrayUpperBound;
+    }
+    // prioritize local arrayUpperBound
+    if (this.astNode.arrayUpperBound != undefined) {
+      arrayUpperBound = this.astNode.arrayUpperBound;
+    }
+    if (typeof arrayUpperBound === "object") {
+      arrayUpperBound = this.getConstantNode(arrayUpperBound.name).value as number;
```

Comment:
weird typeof and `as` here – could this be `!= undefined` instead?

Reply:
moved this logic into `resolvePossibleNumericConstantUsage` (wordy i know ). No longer need casting